lib/hooks.ml

```diff
@@ -0,0 +1,18 @@
+let use_ref initial =
+  let state, _ = Core.use_state (fun () -> ref initial) in
+  state
+
+let use_effect ~on ?(release = Fun.const ()) acquire =
```

Comment:
I'm not entirely convinced by the terminology either, but I still think I prefer the choices I made (shocking, I know!)

`dependencies`, or variations thereof, seems a bit vague in what the dependency relationship actually is, whereas `on` quite clearly alludes to the effect happening "on" those values changing. I think you'll still have to read the documentation to fully understand how this works, but once you've done so it seems to me that `on` will be easier to recall when both reading and writing the code. Also, `dependencies` being such a long word with several possible abbreviations can make it especially difficult to recall the right one.

Another alternative is `watch`, but I'm even less fond of that, despite it being fairly short. I think because it's even more vague.

I see where you're coming from with wanting `cleanup` being more general, but I'm not sure there is actually much of a need for this generality. Even the React documentation you refer to only mentions the need for resource cleanup:

> Often, effects create resources that need to be cleaned up before the component leaves the screen, such as a subscription or timer ID. To do this, the function passed to useEffect may return a clean-up function.

But, perhaps more importantly, there's a subtle difference in this API where instead of returning a cleanup function from the effect, the `acquire` function can return a "resource", which will then  be passed to the `release` function when it's time for cleanup. I think the current naming is much more helpful in suggesting this intended use.

Also keep in mind that it's still possible to pass a generic `unit -> unit` cleanup function to `release`. The difference is just that the naming focuses on (what I think is) the much more common use case.

Reply:
I split `use_effect` into `use_effect`, which has a `cleanup` parameter and no state passed between the effect and cleanup functions, and `use_resource` which behaves like before but makes its purpose even more apparent. While `use_effect` is technically just a thin coat of paint over `use_resource`, I think the distinction makes sense for pedagogical purposes and hope this addresses your concerns as well. Thanks for pointing it out!

I'm still using `~on` for the dependencies though, while not perfect, is still better than the alternatives I think.